### PR TITLE
Support for adding event attributes via the API

### LINKED ISSUE #2210: Add support for modifying event attributes via the API
**Is your feature request related to a problem? Please describe.**
Currently the ability to modify event attributes is only exposed to analyzers and isn't available through the API (not to be confused with sketch attributes which can be modified through the API).

It would be useful to modify event attributes via the API when enriching events with other data sources.

**Describe the solution you'd like**

- Add an EventModifyResource or similar in [timesketch/api/v1/resources/event.py](https://github.com/google/timesketch/blob/master/timesketch/api/v1/resources/event.py#L95)
- Add a route in [timesketch/api/v1/routes.py](https://github.com/google/timesketch/blob/master/timesketch/api/v1/routes.py)
- Add functionality to call this endpoint to the api_client

**Describe alternatives you've considered**
None at this stage.

### PR DESCRIPTION
**Pull request by jonathan-greig**
Fixes #2210

+ Add support for modifying event attributes via the API

**Checks**

- [x] All tests succeed.
- [x] Unit tests added.
- [x] e2e tests added.
- [x] Documentation updated.

### CODE REVIEW

**Comment by berggren** [comment]
Great! I'll keep an eye on the draft status and will review asap

**Comment by jonathan-greig** [comment]
Changing back to draft - have a few changes to make.

File: .gitignore [hunk 1]
```diff
@@ -15,10 +15,10 @@
 /.eggs
 /.tox
 /__pycache__
-/build/
+**/build/
```
Comment (jaegeral) [inline .gitignore:18]: I suggest if there are changes needed to the .gitignore to keep them in a separate PR.

File: api_client/python/timesketch_api_client/sketch_test.py [hunk 2]
```diff
@@ -55,3 +55,25 @@ def test_get_event(self):
         self.assertIsInstance(event_data, dict)
         self.assertTrue('meta' in event_data)
         self.assertTrue('comments' in event_data['meta'])
+
+    def test_add_event_attributes(self):
```
Comment (jaegeral) [inline api_client/python/timesketch_api_client/sketch_test.py:59]: Could you also add a test case for `("Events need to be a list.")` ?
And another test case what happens if you add an attribute that is already present in the event?

File: docs/developers/api-client.md [hunk 3]
```diff
@@ -565,6 +565,30 @@ Both will give you something like:
...
+# event_id and searchindex_id can be found by searching for the target event.
+
+event = [
+  {
+    '_id': event_id,
+    '_index': searchindex_id,
+    '_type': '_doc',
+    'attributes': [
+      {
+        'attr_name': 'A new attribute name',
```
Comment (jaegeral) [inline docs/developers/api-client.md:582]: I wonder if we should encourage people to use whitespaces in attribute names.... WDYT?

File: end_to_end_tests/client_test.py [hunk 4]
```diff
@@ -142,5 +145,49 @@ def test_get_sigma_rule(self):
...
+                    "attributes_added": 1,
+                    "chunks_per_index": {old_event["_index"]: 1},
+                    "error_count": 0,
+                    "errors": [],
+                    "events_modified": 1
+                },
+                "objects": []
+            }
+        )
+        self.assertions.assertIn("foo", new_event["objects"])
```
Comment (jaegeral) [inline end_to_end_tests/client_test.py:190]: Could you also add a test case for `INVALID_ATTRIBUTES_NAMES`?

File: timesketch/lib/testlib.py [hunk 5]
```diff
@@ -65,33 +65,66 @@ def __init__(self):
...
-        """Mock a client search, used for aggregations."""
-        meta = {
-            "es_time": 23,
-            "es_total_count": 5621,
-            "timed_out": False,
-            "max_score": 0.0,
+    def search(self, index, body, size=0, search_type=None):  # pylint: disable=unused-argument
+        """Mock a client search.
+
+        Used for testing both aggregations and adding event attributes.
```
Comment (jaegeral) [inline timesketch/lib/testlib.py:71]: Is there a specific reason you are adding the test cases for adding attributes in that method vs running a separate `_add_event_attribute` method in `BaseTest`?

**Review by jaegeral — COMMENTED** [review]
Some mostly test related comments

File: .gitignore (hunk 1 above)
Comment (jonathan-greig) [reply .gitignore:18]: Moved .gitignore changes to https://github.com/google/timesketch/pull/2311

File: api_client/python/timesketch_api_client/sketch_test.py (hunk 2 above)
Comment (jonathan-greig) [reply api_client/python/timesketch_api_client/sketch_test.py:59]: Added a test case for `("Events need to be a list.")` but I don't think adding one for `add an attribute that is already present in the event` will really be testing anything useful. 

Duplicate attributes won't be added and will just be noted in the `errors` field in the response which is returned directly by `timesketch_api_client.sketch.Sketch.add_event_attributes`.

File: timesketch/lib/testlib.py (hunk 5 above)
Comment (jonathan-greig) [reply timesketch/lib/testlib.py:93]: Not sure I understand exactly, the endpoint makes calls to the opensearch client similar to `EventTaggingResource` so I've added a relevant mock response to `MockOpenSearchClient`. I don't see how adding an `_add_event_attribute` method to `BaseTest` would achieve the same thing but I may be misunderstanding.

Comment (jaegeral) [reply timesketch/lib/testlib.py:93]: ACK

File: timesketch/api/v1/resources/event.py [hunk 6]
```diff
@@ -327,6 +327,201 @@ def get(self, sketch_id):
...
+            A dict with searchindex IDs as keys and the associated events as
+                values.
+        """
+        if not flask_request.is_json:
+            abort(
+                HTTP_STATUS_CODE_BAD_REQUEST,
+                "Request must be in JSON format."
+            )
+        form = flask_request.json
+        events = form.get("events")
```
Comment (jaegeral) [inline timesketch/api/v1/resources/event.py:358]: how about `events = flask_request.json.get...` directly since you are not using `form` anywhere else?

File: timesketch/api/v1/resources/event.py [hunk 7]
```diff
@@ -327,6 +327,201 @@ def get(self, sketch_id):
...
+
+            if event["_index"] not in events_by_index:
+                events_by_index[event["_index"]] = []
+            events_by_index[event["_index"]].append(event)
+
+        return events_by_index
+
+    @login_required
+    def post(self, sketch_id):
+        """Handles POST requests to the resource.
```
Comment (jaegeral) [inline timesketch/api/v1/resources/event.py:412]: Can you make that a little more verbose on what the method will do?

File: timesketch/api/v1/resources/event.py [hunk 8]
```diff
@@ -327,6 +327,201 @@ def get(self, sketch_id):
...
+
+        Args:
+            sketch_id: Integer primary key for a sketch database model
```
Comment (jaegeral) [inline timesketch/api/v1/resources/event.py:415]: `Integer primary key for a sketch database model.`

File: timesketch/api/v1/resources/event.py [hunk 9]
```diff
@@ -327,6 +327,201 @@ def get(self, sketch_id):
...
+
+        Returns:
+            An annotation in JSON (instance of flask.wrappers.Response)
```
Comment (jaegeral) [inline timesketch/api/v1/resources/event.py:418]: `An annotation in JSON (instance of flask.wrappers.Response).`

File: timesketch/api/v1/resources_test.py [hunk 10]
```diff
@@ -323,6 +323,282 @@ def test_invalid_index(self):
...
+        """Test that a content-type other than application/json is handled."""
+        self.login()
+        response = self.client.post(self.resource_url)
+        self.assertEqual(400, response.status_code)
+        self.assertIn(b"Request must be in JSON format.", response.data)
+
+    @mock.patch(
+        "timesketch.api.v1.resources.OpenSearchDataStore", MockDataStore
+    )
+    def test_events_json_validation(self):
```
Comment (jaegeral) [inline timesketch/api/v1/resources_test.py:373]: Please split up those collection of test cases in multiple individual tests

File: timesketch/api/v1/resources_test.py [hunk 11]
```diff
@@ -323,6 +323,282 @@ def test_invalid_index(self):
...
+            }
+        )
+        self.assertEqual(
+            {"1":1, "2":1}, response.json["meta"]["chunks_per_index"])
+
+    @mock.patch(
+        "timesketch.api.v1.resources.OpenSearchDataStore", MockDataStore
+    )
+    def test_add_invalid_attributes(self):
+        """Tests existing attributes cannot be overidden."""
```
Comment (jaegeral) [inline timesketch/api/v1/resources_test.py:532]: again rather split it up in smaller individual tests

**Review by jaegeral — CHANGES_REQUESTED** [review]
PTAL

File: timesketch/api/v1/resources/event.py (hunk 6 above)
Comment (jonathan-greig) [reply timesketch/api/v1/resources/event.py:358]: Done

File: timesketch/api/v1/resources/event.py (hunk 8 above)
Comment (jonathan-greig) [reply timesketch/api/v1/resources/event.py:415]: Done

File: timesketch/api/v1/resources/event.py (hunk 9 above)
Comment (jonathan-greig) [reply timesketch/api/v1/resources/event.py:418]: Done

File: timesketch/api/v1/resources_test.py (hunk 11 above)
Comment (jonathan-greig) [reply timesketch/api/v1/resources_test.py:574]: Done

**Comment by jaegeral** [comment]
LGTM, but to @berggren for the overall impact on the project.

**Review by jaegeral — APPROVED** [review]
LGTM

File: timesketch/api/v1/resources/event.py [hunk 12]
```diff
@@ -327,6 +327,201 @@ def get(self, sketch_id):
         return jsonify(schema)
 
 
+class EventAddAttributeResource(resources.ResourceMixin, Resource):
+    """Resource to add attributes to events."""
+
+    EVENT_FIELDS = ["_id", "_type", "_index", "attributes"]
+    ATTRIBUTE_FIELDS = ["attr_name", "attr_value"]
+    INVALID_ATTRIBUTES_NAMES = [
```
Comment (berggren) [inline timesketch/api/v1/resources/event.py:335]: What about calling it `RESERVED_ATTRIBUTE_NAMES`?

**Review by berggren — CHANGES_REQUESTED** [review]
PTAL

File: end_to_end_tests/client_test.py [hunk 13]
```diff
@@ -190,5 +193,92 @@ def test_get_sigmarule(self):
         count = len(data_frame)
         self.assertions.assertEqual(count, 1)
 
+    def test_add_event_attributes(self):
+        """Tests adding attributes to an event."""
+        sketch = self.api.create_sketch(name="Add event attributes test")
+        sketch.add_event("event message", "2020-01-01T00:00:00", "timestamp_desc")
+        # Wait for new timeline and event to be created
+        time.sleep(1)
```
Comment (berggren) [inline end_to_end_tests/client_test.py:201]: This looks like it can create flaky tests. Could we check that the timeline is ready by looking at the status of it? IF not easily done, we could afford to increase the time at least, say ~5s to make sure the datastore is ready.

**Review by berggren — CHANGES_REQUESTED** [review]
LG, one small thing left.

File: end_to_end_tests/client_test.py (hunk 13 above)
Comment (jonathan-greig) [reply end_to_end_tests/client_test.py]: Added a retry loop to check the status.

**Review by berggren — APPROVED** [review]
LGTM
